### PR TITLE
feat: `slice`, `rawSlice`, `ascii` and `crc32` built-in functions

### LINKED ISSUE #444: Introduce comptime builtins corresponding to FunC string literal types
like `h`, `c`, etc. See https://docs.ton.org/develop/func/literals_identifiers.

I don't think we should extend the syntax of string literals with these suffixes, because those are too cryptic for beginners, we should just introduce those as compile-time builtin functions, like `crc32("foobar")` as a Tact analogue for `"foobar"c` in FunC.

- [x] `"string"s` — defines a raw slice const by its contents (hex-encoded and optionally bit-padded) [added in PR [#787](https://github.com/tact-lang/tact/pull/787) - `rawSlice("string")`];
- [x] `"string"a` — creates a slice const containing MsgAddressInt structure from a specified address [already have it: `address("string")`];
- [x] `"string"u` — creates an int const that corresponds to the hex values of the provided ASCII string [added in PR [#787](https://github.com/tact-lang/tact/pull/787) - `ascii("string")`];
- [x] `"string"h` — creates an int const that is the first 32 bits of the SHA256 hash of the string [no need: can be solved with `sha256` and a bit mask, this will be optimized in compile-time];
- [x] `"string"H` — creates an int const that is all 256 bits of the SHA256 hash of the string [already have it: `sha256("string")`];
- [x] `"string"c` — creates an int const that is crc32 value of the string [added in PR [#787](https://github.com/tact-lang/tact/pull/787): `crc32("string")`].

### PR DESCRIPTION
**Pull request by Gusarich**
## Issue

Closes #444.

* `slice()` isn't actually an equivalent of any string literals, but just a more efficient way to define slice constants from BOC, similarly to `cell()`
* `rawSlice("abcd")` is an equivalent of `"abcd"s` in FunC
* `ascii("hello world")` is an equivalent of `"hello world"u` in FunC
* `crc32("hello world")` is an equivalent of `"hello world"c` in FunC

## Checklist

- [x] I have updated CHANGELOG.md
- [ ] I have documented my contribution in Tact Docs: https://github.com/tact-lang/tact-docs/pull/PR-NUMBER
- [x] I have added tests to demonstrate the contribution is correctly implemented: this usually includes both positive and negative tests, showing the happy path(s) and featuring intentionally broken cases
- [x] I have run all the tests locally and no test failure was reported
- [x] I have run the linter, formatter and spellchecker
- [x] I did not do unrelated and/or undiscussed refactorings

### CODE REVIEW

**Review by anton-trunov — CHANGES_REQUESTED** [review]
1. What does `slice` and `rawSlice` correspond to in FunC string suffixes?
2. Let's add some negative tests, for instance, `ascii("000000000000000000000000000000000")` is supposed to fail due to overflow (33 zeros * 8 = 264 bits)`

File: src/test/e2e-emulated/contracts/intrinsics.tact [hunk 1]
```diff
@@ -6,6 +6,10 @@ contract IntrinsicsTester {
     d: Cell = cell("te6cckEBAQEADgAAGEhlbGxvIHdvcmxkIXgtxbw=");
     e: Int = pow(2, 9);
     f: Int = sha256("hello world");
+    g: Slice = slice("te6cckEBAQEADgAAGEhlbGxvIHdvcmxkIXgtxbw=");
+    h: Slice = rawSlice("hello world");
+    i: Int = ascii("hello world");
+    j: Int = crc32("hello world");
```
Comment (anton-trunov) [inline src/test/e2e-emulated/contracts/intrinsics.tact]: Let's reuse the test cases from the TON docs (https://docs.ton.org/develop/func/literals_identifiers#string-literals). For instance:

> `"transfer(slice, int)"c` becomes 2235694568 int const

**Comment by anton-trunov** [comment]
cspell

**Comment by Gusarich** [comment]
fixed

File: src/test/compilation-failed/const-eval-failed.spec.ts [hunk 2]
```diff
@@ -163,4 +163,9 @@ describe("fail-const-eval", () => {
         errorMessage:
             "Cannot evaluate expression to a constant: repeat argument must be a number between -2^256 (inclusive) and 2^31 - 1 (inclusive)",
     });
+    itShouldNotCompile({
```
Comment (anton-trunov) [inline src/test/compilation-failed/const-eval-failed.spec.ts:166]: Let's also add a negative test with `ascii("")`

Comment (Gusarich) [reply src/test/compilation-failed/const-eval-failed.spec.ts:166]: and same for `crc32` too?

Comment (anton-trunov) [reply src/test/compilation-failed/const-eval-failed.spec.ts:166]: I just checked: in Func `""c` results in `0 PUSHINT`

Comment (anton-trunov) [reply src/test/compilation-failed/const-eval-failed.spec.ts:166]: btw, `ascii("⚡⚡⚡⚡⚡⚡⚡⚡⚡⚡⚡")` should result in overflow, but `ascii("⚡⚡⚡⚡⚡⚡⚡⚡⚡⚡")` should be equal `1563963554659859369353828835329962428465513941646011501275668087180532385` (let's add these two tests too)

File: src/test/e2e-emulated/contracts/intrinsics.tact [hunk 3]
```diff
@@ -6,6 +6,10 @@ contract IntrinsicsTester {
     d: Cell = cell("te6cckEBAQEADgAAGEhlbGxvIHdvcmxkIXgtxbw=");
     e: Int = pow(2, 9);
     f: Int = sha256("hello world");
+    g: Slice = slice("te6cckEBAQEADgAAGEhlbGxvIHdvcmxkIXgtxbw=");
+    h: Slice = rawSlice("abcdef");
```
Comment (anton-trunov) [inline src/test/e2e-emulated/contracts/intrinsics.tact:10]: Let's also have a test with `rawSlice("")`

**Review by anton-trunov — COMMENTED** [review]
Let's add a negative test with `rawSlice(s)` where `s` won't fit into a cell

File: src/interpreter.ts [hunk 4]
```diff
@@ -1133,7 +1133,8 @@ export class Interpreter {
                         this.interpretExpression(ast.args[0]!),
                         ast.args[0]!.loc,
                     );
-                    return BigInt(crc32.str(str));
+                    const c = BigInt(crc32.str(str));
+                    return c < 0 ? c + 4294967296n : c;
```
Comment (anton-trunov) [inline src/interpreter.ts]: https://github.com/SheetJS/js-crc32?tab=readme-ov-file#signed-integers:

> Unconventional for a CRC32 checksum, this library uses signed 32-bit integers. This is for performance reasons. 

> `x >>> 0` converts a number value to unsigned 32-bit integer.

Alternatively, we could use AWS's implementation of CRC-32 which seems to use unsigned numbers: https://www.npmjs.com/package/@aws-crypto/crc32.

Comment (anton-trunov) [reply src/interpreter.ts]: If you decide to keep js-crc32, then let's switch to `x >>> 0` and add this comment about signed numbers, otherwise it's a bit unclear

File: src/test/compilation-failed/contracts/const-eval-rawslice-overflow.tact [hunk 5]
```diff
@@ -0,0 +1,5 @@
+contract AsciiOverflow {
+    get fun getRawSlice_fail(): Slice {
+        return rawSlice("abcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcd");
```
Comment (anton-trunov) [inline src/test/compilation-failed/contracts/const-eval-rawslice-overflow.tact:3]: weirdly enough, FunC does not complain about this particular string:
```func
#pragma version =0.4.4;
#pragma allow-post-modification;
#pragma compute-asm-ltr;

_ %foo() method_id(0xff) {
    return "abcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcd"s;
}
```

the code above compiles and produces

```
❯ ./func test.fc
// automatically generated from `test.fc`
255 DECLMETHOD %foo
%foo PROC:<{
  x{abcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcd} PUSHSLICE
}>
```

but if I add one more hex character to the string, then I get an error:

```
test.fc:6:12: error: Invalid hex bitstring constant `abcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcda`
      return "abcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabcdabc ...
             ^
```

Comment (Gusarich) [reply src/test/compilation-failed/contracts/const-eval-rawslice-overflow.tact:3]: seems like a FunC problem. when I run the generated fift output - it crashes.

Comment (anton-trunov) [reply src/test/compilation-failed/contracts/const-eval-rawslice-overflow.tact:3]: Oh, interesting. Can you please file an issue to the monorepo?

File: src/interpreter.ts (hunk 4 above)
Comment (Gusarich) [reply src/interpreter.ts]: done

File: src/test/compilation-failed/contracts/const-eval-rawslice-overflow.tact (hunk 5 above)
Comment (Gusarich) [reply src/test/compilation-failed/contracts/const-eval-rawslice-overflow.tact:3]: https://github.com/ton-blockchain/ton/issues/1153

**Review by anton-trunov — APPROVED** [review]
Great stuff!
